src/type-definitions/plainscript-types.ts

```diff
@@ -0,0 +1,40 @@
+
+/**
+ * Functions as a namespace for type aliases and interfaces
+ * that are used commonly throughout the ast files.
+ */
+
+import AssignmentStatement from '../ast/assignment-statement';
+import BinaryExpression from '../ast/binary-expression';
+import BooleanLiteral from '../ast/boolean-literal';
+import BreakStatement from '../ast/break-statement';
+import Call from '../ast/call';
+import FunctionObject from '../ast/function-object';
+import IdentifierExpression from '../ast/identifier-expression';
+import ListExpression from '../ast/list-expression';
+import NumericLiteral from '../ast/numeric-literal';
+import Parameter from '../ast/parameter';
+import ReturnStatement from '../ast/return-statement';
+import StringLiteral from '../ast/string-literal';
+import SubscriptedExpression from '../ast/subscripted-expression';
+import UnaryExpression from '../ast/unary-expression';
+import Variable from '../ast/variable';
+import VariableDeclaration from '../ast/variable-declaration';
+import Context from '../semantics/context';
+
+export type Literal = NumericLiteral | StringLiteral | BooleanLiteral;
+export type Referent = FunctionObject | Parameter | Variable;
+
+export type Expression = BinaryExpression | UnaryExpression | Literal | ListExpression
+    | Call | SubscriptedExpression | IdentifierExpression;
+
+export type Statement = AssignmentStatement | BreakStatement | ReturnStatement | Call
+    | BreakStatement | VariableDeclaration;
+
+export type Body = Statement[];
+
+
+export interface IAstNode<T> {
+    analyze(context?: Context): void;
+    optimize?(): T | Literal | void | null;
+}
```

Comment:
I'm wondering if this could be a base class and define `gen` here so you don't need all the empty `gen` declarations in each of the classes. I know that kind of implementation inheritance sucks tho. I dunno. I wish TS had default methods in interfaces like Java. 😱 Maybe mixins could do it?

Reply:
I think both class inheritance and mixins could solve getting rid of these empty gen stub methods in the ast classes. I think you could argue for either but I think class inheritance might make more sense semantically here, since we actually do want all ast node classes to have an empty gen method to satisfy TypeScript's checks in the generator whereas mixins seems to be meant for giving some classes some resuable implemented but not all of the classes.

Reply:
Makes sense. Inheritance might actually work well actually. All of those nodes _are_ nodes; this is more accurate than saying they all "behave like nodes." While you are at it, you can make things like Statement, Expression, etc. be actually classes too (albeit abstract). Also you can put in some default optimize methods, e.g. where optimize has nothing to do and just returns itself.

Reply:
Although inheritance will require me to throw in a `super();` in the constructors of all the ast classes that inherit from a base class. But oh well, can't win them all

Reply:
It does, but in may cases it will make sense because there _are_ things to pass to the superclass constructor: calling `super(x)` is the same number of lines as a `this.x = x`. Well I hope there are a lot of those.....

Reply:
Would a type alias be just as appropriate for Statements, Expressions, etc ? Most of the statement classes share no common fields actually ! So the abstract class would just be an empty class, adding extra verboseness

Reply:
It would be empty yes but if you look at the class hierarchy diagram it makes a wonderful partitioning. ALSO, you can write things like `Statement[]` and it will apply to all statements and not allow expressions and programs and other things like that.

Reply:
@rtoal ok pushed updates to group some of the classes under common abstract classes !

Things to note:
- put these classes in an abstract folder within the ast folder as they do not even get matched up by the parser
- left `Body` as a type alias for `Statement[]` in the type-definition file (can change this if recommended into an abstract class or just plain convert all references to `Statement[]`)